libpromises/mod_custom.c

```diff
@@ -0,0 +1,741 @@
+    {
+        module->message = JsonObjectCreate(10);
+    }
+
+    JsonElement *attributes = JsonObjectGet(module->message, "attributes");
+    if (attributes == NULL)
+    {
+        attributes = JsonObjectCreate(10);
+        JsonObjectAppendObject(module->message, "attributes", attributes);
+    }
+
+    JsonObjectAppendString(attributes, key, value);
+}
+
+static void PromiseModule_Send(PromiseModule *module)
+{
+    assert(module != NULL);
+
+    if (module->json)
+    {
+        Writer *w = FileWriter(module->input);
+        JsonWriteCompact(w, module->message);
+        FileWriterDetach(w);
+        DESTROY_AND_NULL(JsonDestroy, module->message);
+        fprintf(module->input, "\n\n");
+        fflush(module->input);
+        return;
+    }
+
+    Seq *message = SeqNew(10, free);
+
+    JsonIterator iter = JsonIteratorInit(module->message);
+    const char *key;
+    while ((key = JsonIteratorNextKey(&iter)) != NULL)
+    {
+        if (StringEqual("attributes", key))
+        {
+            JsonElement *attributes = JsonIteratorCurrentValue(&iter);
+            JsonIterator attr_iter = JsonIteratorInit(attributes);
+
+            const char *attr_name;
+            while ((attr_name = JsonIteratorNextKey(&attr_iter)) != NULL)
+            {
+                const char *attr_val = JsonPrimitiveGetAsString(
+                    JsonIteratorCurrentValue(&attr_iter));
+                char *attr_line = NULL;
+                xasprintf(&attr_line, "attribute_%s=%s", attr_name, attr_val);
+                SeqAppend(message, attr_line);
+            }
+        }
+        else
+        {
+            const char *value =
+                JsonPrimitiveGetAsString(JsonIteratorCurrentValue(&iter));
+            char *line = NULL;
+            xasprintf(&line, "%s=%s", key, value);
+            SeqAppend(message, line);
+        }
+    }
+
+    PromiseModule_SendMessage(module, message);
+    SeqDestroy(message);
+}
+
+static void PromiseModule_AppendAllAttributes(
+    PromiseModule *module, const Promise *pp)
+{
+    assert(module != NULL);
+    assert(pp != NULL);
+
+    const size_t attributes = SeqLength(pp->conlist);
+    for (size_t i = 0; i < attributes; i++)
+    {
+        const Constraint *attribute = SeqAt(pp->conlist, i);
+        const char *const name = attribute->lval;
+        char *const value = RvalToString(attribute->rval);
+        PromiseModule_AppendAttribute(module, name, value);
+        free(value);
+    }
+}
+
+static inline bool StringContainsUnresolved(const char *str)
+{
+    return (strchr(str, '$') != NULL);
+}
+
+static inline bool CustomPromise_IsFullyResolved(const Promise *pp)
+{
+    assert(pp != NULL);
+
+    if (StringContainsUnresolved(pp->promiser))
+    {
+        return false;
+    }
+    const size_t attributes = SeqLength(pp->conlist);
+    for (size_t i = 0; i < attributes; i++)
+    {
+        const Constraint *attribute = SeqAt(pp->conlist, i);
+        if (attribute->rval.type != RVAL_TYPE_SCALAR)
```

Comment:
Do all the custom promises have to accept only scalars?

Reply:
For this iteration, yes.